examples/coap-server.c

```diff
@@ -377,9 +380,196 @@ check_async(coap_context_t *ctx,
+      else {
+        coap_log(LOG_WARNING,
+                 "No cache entry available for the non-first BLOCK1\n");
+      }
+      response->code = COAP_RESPONSE_CODE(500);
+      return;
+    }
+
+    data_so_far = coap_cache_get_app_data(cache_entry);
+    if (offset == 0) {
+      if (data_so_far) {
+        coap_delete_binary(data_so_far);
+        data_so_far = NULL;
+      }
+    }
+    else if (offset >
+             (data_so_far ? data_so_far->length : 0)) {
+      /* Upload is not sequential - block missing */
+      response->code = COAP_RESPONSE_CODE(408);
+      return;
+    }
+    else if (offset <
+             (data_so_far ? data_so_far->length : 0)) {
+      /* Upload is not sequential - block duplicated */
+      goto just_respond;
+    }
+
+    if (coap_get_data(request, &size, &data) && (size > 0)) {
+      if (!data_so_far) {
+        data_so_far = coap_new_binary(size);
+        memcpy(data_so_far->s, data, size);
+      }
+      else {
+        /* Add in new block to end of current data */
+        data_so_far = coap_resize_binary(data_so_far, offset + size);
+        memcpy(&data_so_far->s[offset], data, size);
+      }
+    }
+    else if (!block1.m && block1.num == 0) {
+      /* Empty first and only block */
+      if (data_so_far) {
+        coap_delete_binary(example_data_ptr);
+      }
+      data_so_far = coap_new_binary(0);
+    }
+
+    if (!block1.m) {
+      /* all the data in - now update the resource */
+      coap_delete_binary(example_data_ptr);
+      example_data_ptr = data_so_far;
+      coap_cache_set_app_data(cache_entry, NULL, NULL);
+      if ((option = coap_check_option(request, COAP_OPTION_CONTENT_FORMAT,
+                                      &opt_iter)) != NULL) {
+        example_data_media_type =
+            coap_decode_var_bytes (coap_opt_value (option),
+                                   coap_opt_length (option));
+      }
+      else {
+        example_data_media_type = COAP_MEDIATYPE_TEXT_PLAIN;
+      }
+      coap_resource_notify_observers(resource, NULL);
+    }
+    else {
+      /* save the updated data for the next block */
+      coap_cache_set_app_data(cache_entry,data_so_far, cache_free_app_data);
+    }
+
+just_respond:
+    if (block1.m) {
+      response->code = COAP_RESPONSE_CODE(231);
+    }
+    else {
+      response->code = COAP_RESPONSE_CODE(204);
+    }
+    coap_add_option(response,
+                    COAP_OPTION_BLOCK1,
+                    coap_encode_var_safe(buf, sizeof(buf),
+                                         ((block1.num << 4) |
+                                          (block1.m << 3) |
+                                          block1.szx)),
+                    buf);
+  }
+  else if (coap_get_data(request, &size, &data) && (size > 0)) {
+    /* Not a BLOCK1 with data */
+    if (example_data_ptr) {
+      coap_delete_binary(example_data_ptr);
+    }
+    example_data_ptr = coap_new_binary(size);
+    memcpy (example_data_ptr->s, data, size);
```

Comment:
Need to check `example_data_ptr` before `memcpy()`

Reply:
Fixed